aws/data_source_aws_ec2_transit_gateway_peering_attachment.go

```diff
@@ -0,0 +1,96 @@
+package aws
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+)
+
+func dataSourceAwsEc2TransitGatewayPeeringAttachment() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsEc2TransitGatewayPeeringAttachmentRead,
+
+		Schema: map[string]*schema.Schema{
+			"filter": ec2CustomFiltersSchema(),
+			"id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"peer_account_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"peer_region": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"peer_transit_gateway_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"tags": tagsSchemaComputed(),
+			"transit_gateway_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceAwsEc2TransitGatewayPeeringAttachmentRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	input := &ec2.DescribeTransitGatewayPeeringAttachmentsInput{}
+
+	if v, ok := d.GetOk("id"); ok {
+		input.TransitGatewayAttachmentIds = aws.StringSlice([]string{v.(string)})
+	}
+
+	input.Filters = buildEC2CustomFilterList(d.Get("filter").(*schema.Set))
+	if v := d.Get("tags").(map[string]interface{}); len(v) > 0 {
+		input.Filters = append(input.Filters, ec2TagFiltersFromMap(v)...)
+	}
+	if len(input.Filters) == 0 {
+		// Don't send an empty filters list; the EC2 API won't accept it.
+		input.Filters = nil
+	}
+
+	log.Printf("[DEBUG] Reading EC2 Transit Gateway Peering Attachments: %s", input)
+	output, err := conn.DescribeTransitGatewayPeeringAttachments(input)
+
+	if err != nil {
+		return fmt.Errorf("error reading EC2 Transit Gateway Peering Attachments: %s", err)
+	}
+
+	if output == nil || len(output.TransitGatewayPeeringAttachments) == 0 {
+		return errors.New("error reading EC2 Transit Gateway Peering Attachment: no results found")
+	}
+
+	if len(output.TransitGatewayPeeringAttachments) > 1 {
+		return errors.New("error reading EC2 Transit Gateway Peering Attachment: multiple results found, try adjusting search criteria")
+	}
+
+	transitGatewayPeeringAttachment := output.TransitGatewayPeeringAttachments[0]
+
+	if transitGatewayPeeringAttachment == nil {
+		return errors.New("error reading EC2 Transit Gateway Peering Attachment: empty result")
+	}
+
+	d.Set("peer_account_id", transitGatewayPeeringAttachment.AccepterTgwInfo.OwnerId)
+	d.Set("peer_region", transitGatewayPeeringAttachment.AccepterTgwInfo.Region)
+	d.Set("peer_transit_gateway_id", transitGatewayPeeringAttachment.AccepterTgwInfo.TransitGatewayId)
+	d.Set("transit_gateway_id", transitGatewayPeeringAttachment.RequesterTgwInfo.TransitGatewayId)
```

Comment:
I wonder if this information should be set based on the current account ID and region of the provider? This data source could be used on either side of the attachment, so we'll want it to reflect the correct peer versus local.

```suggestion
	local := transitGatewayPeeringAttachment.RequesterTgwInfo
	peer := transitGatewayPeeringAttachment.AccepterTgwInfo
	
	if aws.StringValue(transitGatewayPeeringAttachment.AccepterTgwInfo.OwnerId) == meta.(*AWSClient).accountid && aws.StringValue(transitGatewayPeeringAttachment.AccepterTgwInfo.Region) == meta.(*AWSClient).region {
		local = transitGatewayPeeringAttachment.AccepterTgwInfo
		peer = transitGatewayPeeringAttachment.RequesterTgwInfo
	}

	d.Set("peer_account_id", peer.OwnerId)
	d.Set("peer_region", peer.Region)
	d.Set("peer_transit_gateway_id", peer.TransitGatewayId)
	d.Set("transit_gateway_id", local.TransitGatewayId)
```

Another (I think less desirable choice) might be to expose it directly as accepter/requester attributes too.

We may also want to consider a separate test where the `aws_ec2_transit_gateway_peering_attachment` is using `provider = aws.alternate` to test this scenario. 👍

Reply:
Hi @bflad, thanks for the suggestion, this is a nicer way of doing it.

Will write tests for this scenario shortly.

Reply:
@bflad 9221ed7 test scenario added for data source using `provider = aws.alternate`

verified acceptance test:

```go
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_Filter_differentAccount'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_Filter_differentAccount -timeout 120m
=== RUN   TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_Filter_differentAccount
=== PAUSE TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_Filter_differentAccount
=== CONT  TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_Filter_differentAccount
--- PASS: TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_Filter_differentAccount (316.25s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       317.401s

$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_ID_differentAccount'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_ID_differentAccount -timeout 120m
=== RUN   TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_ID_differentAccount
=== PAUSE TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_ID_differentAccount
=== CONT  TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_ID_differentAccount
--- PASS: TestAccAWSEc2TransitGatewayPeeringAttachmentDataSource_ID_differentAccount (304.40s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       305.575s
```